File: GardenSystem/Logger/Logger/Logger/Log.vb
Imports System.IO
Public Class Log

    Private LogName As String
    Private lockobj As New Object

    Sub New(pLogName As String)
        LogName = pLogName
    End Sub

    Private Function DirectoryName() As String
        Dim assembly As System.Reflection.Assembly = Reflection.Assembly.GetExecutingAssembly()
        Dim logPath As String = Path.Combine(Path.GetDirectoryName(assembly.Location), String.Format("Log\{0:0000}\{1:00}\{2:00}\", DateTime.Now.Year, DateTime.Now.Month, DateTime.Now.Day))
        Return logPath
    End Function

    Private Function FileName() As String
        Dim logFile As String = Path.Combine(DirectoryName, String.Format("{0}.txt", LogName))
        Return logFile
    End Function

    Public Sub Add(message As String)
        SyncLock lockobj
            message = message.Replace(vbCrLf, " ")
            Dim log As String = String.Format(DateTime.Now.ToString("yyy/MM/dd HH:mm:ss"))

            If Not Directory.Exists(DirectoryName) Then
                Directory.CreateDirectory(DirectoryName)
            End If

            File.AppendAllText(FileName, log & " " & message & Environment.NewLine)
        End SyncLock
    End Sub

    Public Function Read() As String
        Dim wText As String = String.Empty

        SyncLock lockobj
            If File.Exists(FileName) Then
                Try
                    wText = File.ReadAllText(FileName)
                Catch ex As Exception
                    Throw New Exception("Read - " & ex.Message)
                End Try
            End If
        End SyncLock

        Return wText
    End Function

    Public Function ReadLatest(rows As Int16) As String
        Dim wText As String = String.Empty
        Dim r_splittext() As String

        SyncLock lockobj
            If File.Exists(FileName) And Not (rows = 0) Then
                Try
                    r_splittext = File.ReadAllLines(FileName)
                    If rows >= r_splittext.Length Then
                        rows = r_splittext.Length
                    End If
                    For i As Int16 = r_splittext.Length - 1 To r_splittext.Length - rows Step -1
                        wText += (r_splittext(i) + vbCrLf)
                    Next
                Catch ex As Exception
                    Throw New Exception("ReadLatest - " & ex.Message)
                End Try
            End If
        End SyncLock

        Return wText
    End Function
End Class

File: GardenSystem/LoggerCS/Log.cs
using System;
using System.IO;
using Microsoft.VisualBasic;

namespace LoggerCS
{
    public class Log
    {
        private string LogName;
        private object lockobj = new object();

        public Log(string pLogName)
        {
            LogName = pLogName;
        }

        private string DirectoryName()
        {
            System.Reflection.Assembly assembly = System.Reflection.Assembly.GetExecutingAssembly();
            string logPath = Path.Combine(Path.GetDirectoryName(assembly.Location), string.Format(@"Log\{0:0000}\{1:00}\{2:00}\", DateTime.Now.Year, DateTime.Now.Month, DateTime.Now.Day));
            return logPath;
        }

        private string FileName()
        {
            string logFile = Path.Combine(DirectoryName(), string.Format("{0}.txt", LogName));
            return logFile;
        }

        public void Add(string message)
        {
            lock (lockobj)
            {
                message = message.Replace(Constants.vbCrLf, " ");
                string log = string.Format(DateTime.Now.ToString("yyy/MM/dd HH:mm:ss"));

                if (!Directory.Exists(DirectoryName()))
                    Directory.CreateDirectory(DirectoryName());

                File.AppendAllText(FileName(), log + " " + message + Environment.NewLine);
            }
        }

        public string Read()
        {
            string wText = string.Empty;

            lock (lockobj)
            {
                if (File.Exists(FileName()))
                {
                    try
                    {
                        wText = File.ReadAllText(FileName());
                    }
                    catch (Exception ex)
                    {
                        throw new Exception("[Read] - " + ex.Message);
                    }
                }
            }

            return wText;
        }

        public string ReadLatest(Int16 rows)
        {
            string wText = string.Empty;
            string[] r_splittext;

            lock (lockobj)
            {
                if (File.Exists(FileName()))
                {
                    try
                    {
                        r_splittext = File.ReadAllLines(FileName());
                        //if (rows >= r_splittext.Length)
                        //    rows = r_splittext.Length;
                        for (int i = r_splittext.Length - 1; i >= r_splittext.Length - rows; i += -1)
                            wText += (r_splittext[i] + Constants.vbCrLf);
                    }
                    catch (Exception ex)
                    {
                        throw new Exception("[ReadLatest] - " + ex.Message);
                    }
                }
            }

            return wText;
        }
    }
}
